substrate/client/rpc/src/author/mod.rs

```diff
@@ -123,10 +123,25 @@ where
 
 		runtime_api.register_extension(KeystoreExt::from(self.keystore.clone()));
 
-		runtime_api
-			.generate_session_keys(best_block_hash, None)
-			.map(Into::into)
-			.map_err(|api_err| Error::Client(Box::new(api_err)).into())
+		let version = runtime_api
+			.api_version::<dyn SessionKeys<P::Block>>(best_block_hash)
+			.map_err(|api_err| Error::Client(Box::new(api_err)))?
+			.ok_or_else(|| Error::MissingSessionKeysApi)?;
+
+		let res = if version < 2 {
+			#[allow(deprecated)]
+			runtime_api
+				.generate_session_keys_before_version_2(best_block_hash, None)
+				.map(Into::into)
+				.map_err(|api_err| Error::Client(Box::new(api_err)))
+		} else {
+			runtime_api
+				.generate_session_keys(best_block_hash, Vec::new(), None)
+				.map(|sk| sk.keys.into())
```

Comment:
If I understood correctly, we may want to change this RPC call to return an encoded `GeneratedSessionKeys` right? (Probably with an empty proofs vec if version < 2).

That is, also return the proofs so that the user can eventually submit the `set-keys` extrinsic to the session pallet which now requires the proofs as well.

Otherwise the user should manually (re)generate these proofs before on-chain keys registration

Reply:
> If I understood correctly, we may want to change this RPC call to return an encoded `GeneratedSessionKeys` right?

No. The RPC should not be used anymore. People should just directly call the runtime api using `state_call`.

Reply:
Clear